Function calls in annotation arguments fail in FIR

### Description

Evaluated annotation arguments fail with a compiler error.

```
Caused by: org.jetbrains.kotlin.utils.exceptions.KotlinIllegalArgumentExceptionWithAttachments: Exception in declaration checkers
	at org.jetbrains.kotlin.fir.analysis.checkers.declaration.DeclarationCheckersDiagnosticComponent.visitRegularClass(DeclarationCheckersDiagnosticComponent.kt:289)
	at org.jetbrains.kotlin.fir.analysis.checkers.declaration.DeclarationCheckersDiagnosticComponent.visitRegularClass(DeclarationCheckersDiagnosticComponent.kt:25)
	at org.jetbrains.kotlin.fir.declarations.FirRegularClass.accept(FirRegularClass.kt:49)
	at org.jetbrains.kotlin.fir.analysis.collectors.CheckerRunningDiagnosticCollectorVisitor.checkElement(CheckerRunningDiagnosticCollectorVisitor.kt:24)
	at org.jetbrains.kotlin.fir.analysis.collectors.AbstractDiagnosticCollectorVisitor.access$checkElement(AbstractDiagnosticCollectorVisitor.kt:36)
	at org.jetbrains.kotlin.fir.analysis.collectors.AbstractDiagnosticCollectorVisitor.visitRegularClass(AbstractDiagnosticCollectorVisitor.kt:650)
	at org.jetbrains.kotlin.fir.analysis.collectors.AbstractDiagnosticCollectorVisitor.visitRegularClass(AbstractDiagnosticCollectorVisitor.kt:36)
	at org.jetbrains.kotlin.fir.declarations.FirRegularClass.accept(FirRegularClass.kt:49)
	at org.jetbrains.kotlin.fir.declarations.impl.FirFileImpl.acceptChildren(FirFileImpl.kt:60)
	at org.jetbrains.kotlin.fir.analysis.collectors.AbstractDiagnosticCollectorVisitor.visitNestedElements(AbstractDiagnosticCollectorVisitor.kt:44)
	at org.jetbrains.kotlin.fir.analysis.collectors.AbstractDiagnosticCollectorVisitor.visitFile(AbstractDiagnosticCollectorVisitor.kt:1351)
	at org.jetbrains.kotlin.fir.analysis.collectors.AbstractDiagnosticCollectorVisitor.visitFile(AbstractDiagnosticCollectorVisitor.kt:36)
	at org.jetbrains.kotlin.fir.declarations.FirFile.accept(FirFile.kt:43)
	at org.jetbrains.kotlin.fir.analysis.collectors.AbstractDiagnosticCollector.collectDiagnostics(AbstractDiagnosticCollector.kt:36)
	at org.jetbrains.kotlin.fir.pipeline.AnalyseKt.runCheckers(analyse.kt:37)
	... 38 more
Caused by: java.lang.IllegalStateException: Unexpected annotation argument type: class org.jetbrains.kotlin.fir.expressions.impl.FirFunctionCallImpl - Q|com/example/myapplication/ExampleConstants|.R|com/example/myapplication/ExampleConstants.Companion.CONTROLLER|.R|kotlin/String.plus|(Q|com/example/myapplication/DeepLinkStrategyId|.R|com/example/myapplication/DeepLinkStrategyId.ACTION_DISPLAY|). This is a bug in the Metro compiler, please report it to https://github.com/zacsweers/metro. 
	at dev.zacsweers.metro.compiler.UtilKt.reportCompilerBug(util.kt:209)
	at dev.zacsweers.metro.compiler.fir.FirKt.computeAnnotationHash(fir.kt:380)
	at dev.zacsweers.metro.compiler.fir.MetroFirAnnotation.cachedHashKey_delegate$lambda$0(MetroFirAnnotation.kt:26)
	at kotlin.UnsafeLazyImpl.getValue(Lazy.kt:100)
	at dev.zacsweers.metro.compiler.fir.MetroFirAnnotation.getCachedHashKey(MetroFirAnnotation.kt:26)
	at dev.zacsweers.metro.compiler.fir.MetroFirAnnotation.hashCode(MetroFirAnnotation.kt:42)
	at java.base/java.util.HashMap.hash(Unknown Source)
	at java.base/java.util.HashMap.put(Unknown Sourc
```

### Self-contained Reproducer

Sample project:
https://github.com/agrosner/metro-reproducible-bugs

```kotlin
@Module
@ContributesTo(AppScope::class)
interface ExampleModule {


    @Binds
    @IntoMap
    @StringKey(ExampleConstants.CONTROLLER + DeepLinkStrategyId.ACTION_DISPLAY) <--- here
    fun provideSoftForcedUpgradePreferencesDeepLinkPath(exampleClass: ExampleClass): ExampleClassInterface
}
```

making the string key a regular constant like `@StringKey("something")` compiles fine.

### Metro version

0.7.7

### Context

MacOS, Android platform.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@ Changelog
 - **New**: New `dev.zacsweers.metro:metrox-android` artifact for Android `AppComponentFactory` support. See the [docs](https://zacsweers.github.io/metro/latest/metrox-android) for more information.
 - **Enhancement**: Lazily validate multibindings. Previously, multibindings were validated eagerly even if they were unused in a graph.
 - **Enhancement**: Report all duplicate bindings errors during graph construction rather than failing at first.
+- **Enhancement**: Support `FirFunctionCall` evaluation in annotation arguments during FIR.
 - **Fix**: Catch more `IrErrorType` error types cases and report context/advice where possible.
 - **Fix**: Dedupe binding containers contributed to both parent and child graphs.
 - **Fix**: Fix support for Anvil's `exclude` argument in dependency graph annotations when Anvil interop is enabled.

--- a/compiler-tests/src/test/data/box/dependencygraph/EvaluatableConstantsAreEvaluatedInFir.kt
+++ b/compiler-tests/src/test/data/box/dependencygraph/EvaluatableConstantsAreEvaluatedInFir.kt
@@ -1,0 +1,36 @@
+// https://github.com/ZacSweers/metro/issues/1427
+
+object ExampleConstants {
+  const val CONTROLLER = "controller"
+}
+
+object DeepLinkStrategyId {
+  const val ACTION_DISPLAY = "actiondisplay"
+}
+
+private const val KEY_2 = "${ExampleConstants.CONTROLLER}2${DeepLinkStrategyId.ACTION_DISPLAY}"
+
+@BindingContainer
+@ContributesTo(AppScope::class)
+object ExampleContainer {
+  @Provides
+  @IntoMap
+  @StringKey(ExampleConstants.CONTROLLER + DeepLinkStrategyId.ACTION_DISPLAY)
+  fun provideString(): String = "hello"
+  @Provides
+  @IntoMap
+  @StringKey(KEY_2)
+  fun provideString2(): String = "hello2"
+}
+
+@DependencyGraph(AppScope::class)
+interface AppGraph {
+  val strings: Map<String, String>
+}
+
+fun box(): String {
+  val strings = createGraph<AppGraph>().strings
+  assertEquals("hello", strings[ExampleConstants.CONTROLLER + DeepLinkStrategyId.ACTION_DISPLAY])
+  assertEquals("hello2", strings[KEY_2])
+  return "OK"
+}

--- a/compiler-tests/src/test/java/dev/zacsweers/metro/compiler/BoxTestGenerated.java
+++ b/compiler-tests/src/test/java/dev/zacsweers/metro/compiler/BoxTestGenerated.java
@@ -466,6 +466,12 @@ public class BoxTestGenerated extends AbstractBoxTest {
     }
 
     @Test
+    @TestMetadata("EvaluatableConstantsAreEvaluatedInFir.kt")
+    public void testEvaluatableConstantsAreEvaluatedInFir() {
+      runTest("compiler-tests/src/test/data/box/dependencygraph/EvaluatableConstantsAreEvaluatedInFir.kt");
+    }
+
+    @Test
     @TestMetadata("ExposingIncludedGraphsDirectly.kt")
     public void testExposingIncludedGraphsDirectly() {
       runTest("compiler-tests/src/test/data/box/dependencygraph/ExposingIncludedGraphsDirectly.kt");

--- a/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/fir.kt
+++ b/compiler/src/main/kotlin/dev/zacsweers/metro/compiler/fir/fir.kt
@@ -376,6 +376,12 @@ internal fun FirAnnotationCall.computeAnnotationHash(
               ?.classId
           }
 
+          is FirFunctionCall -> {
+            // This is some constant-able expression like "foo" + "bar" in an annotation arg, which
+            // is legal
+            arg.evaluateAs<FirLiteralExpression>(session)?.value
+          }
+
           else -> {
             reportCompilerBug(
               "Unexpected annotation argument type: ${arg::class.java} - ${arg.render()}"

